torch/csrc/dynamo/guards.cpp

```diff
@@ -574,6 +574,465 @@ static struct PyModuleDef _module = {
+    this->result = result;
+    this->num_guards_executed = num_guards_executed;
+    this->failure_reason = failure_reason;
+  }
+
+  std::string repr() {
+    return "GuardDebugInfo(num_guards_executed=" +
+        std::to_string(num_guards_executed) +
+        ", failure_reason=" + failure_reason + ")";
+  }
+
+  // Whether the guard passed or failed.
+  bool result;
+
+  // Failure reason for a leaf guard.
+  std::string failure_reason;
+
+  // Total number of executed guards so far.
+  int num_guards_executed;
+};
+
+/**
+ * Base class for the terminal or leaf guard in the GuardManager hierarchy.
+ */
+class LeafGuard {
+ public:
+  // check function could be called from python. This is useful for debugging
+  // purpose.
+  bool check(py::handle value) {
+    return check_nopybind(value.ptr());
+  }
+
+  GuardDebugInfo debug_check(py::handle value) {
+    return debug_check_nopybind(value.ptr());
+  }
+
+  GuardDebugInfo debug_check_nopybind(PyObject* value) { // borrowed ref
+    bool result = check_nopybind(value);
+    std::string failure_reason = "";
+    if (!result) {
+      failure_reason = get_failure_reason(value);
+    }
+    return GuardDebugInfo(result, failure_reason, 0);
+  }
+
+  // This is on the hot path and avoids any refcounting code from pybind. This
+  // is not exposed to Python and can only be called from C++.
+  virtual bool check_nopybind(PyObject* value) = 0;
+  virtual std::string get_failure_reason(PyObject* value) = 0;
+  virtual ~LeafGuard() = default;
+};
+
+/**
+ * Represents a leaf guard that accepts the python guard check function. We
+ * would like to have most of the guards in C++ (to avoid a Python function
+ * call).  But, it will take some time to reach that goal. Also, there might be
+ * cases where its too tedious to write an equivalent C++ guard.
+ *
+ * PythonLambdaGuard allows us to gradually move to C++. We can start from all
+ * guards of type PythonLambaGuard and incrementally move expensive guards to
+ * C++.
+ */
+class PythonLambdaGuard : public LeafGuard {
+ public:
+  PythonLambdaGuard(py::object guard_check_fn, py::object print_failure_fn) {
+    if (py::isinstance<py::function>(guard_check_fn) &&
+        py::isinstance<py::function>(print_failure_fn)) {
+      _guard_check_fn = py::cast<py::function>(guard_check_fn);
+      _guard_check_fn_pyobj = _guard_check_fn.ptr();
+      _print_failure_fn = py::cast<py::function>(print_failure_fn);
+    } else {
+      throw py::type_error("PythonLambdaGuard expects callables");
+    }
+  }
+
+  // Runs the lambda function with the current f_locals value.
+  bool check_nopybind(PyObject* value) override { // borrowed ref
+    return PyObject_IsTrue(PyObject_CallOneArg(_guard_check_fn_pyobj, value));
+  }
+
+  std::string get_failure_reason(PyObject* value) override {
+    return py::cast<std::string>(_print_failure_fn(py::handle(value)));
+  }
+
+ private:
+  // The user provided lambda function for check_fn.
+  py::function _guard_check_fn;
```

Comment:
Lost in squashing commits.

@voznesenskym  says "I don't understand the ownership semantics of holding something like this"

Reply:
IIUC, I am using py::object in the constructor, and then copying it in the this py::function, which increases a ref count. I don't think ownership changes. Its just that `LambdaGuard` object holds a reference to the python callables - https://pybind11.readthedocs.io/en/stable/reference.html#with-reference-counting


@jansel correct me if I am wrong.